For reference, this is the new problem: 

```console
$ nix-build -A tests.writers
these derivations will be built:
  /nix/store/hkpg0m44v3kvkgcjhzklbzhy7hyqbw1w-test_writers.drv
  /nix/store/1laarw0v4xnl6jdancfswa9imzddad85-test-writers.drv
  /nix/store/zmzl0phrj8fkzr8zch1xizysyybj7fbs-test_writers.drv
  /nix/store/awn033v3vlj7g31mcznd9caq6cr5vdvq-test-writers.drv
  /nix/store/fvsjrpls1apn2z6qcxz77cyz8z0rdl7c-test_haskell.drv
  /nix/store/xnkdwkv55xqahvqh2rsifasnha5li7rw-test_haskell.drv
  /nix/store/kx2l83gyvryka0x3nlkw8323kgs6qil3-test-writers.drv
  /nix/store/xcn490j6jsgji1mrgsvpvlx3pjjlpzqg-test-writers.drv
  /nix/store/vl2imbpfjhqc8cxqidpk3kfdvvqgj8dy-test-writers.drv
building '/nix/store/fvsjrpls1apn2z6qcxz77cyz8z0rdl7c-test_haskell.drv'...
[1 of 1] Compiling Main             ( tmp.hs, tmp.o )
Linking tmp ...
building '/nix/store/xnkdwkv55xqahvqh2rsifasnha5li7rw-test_haskell.drv'...
[1 of 1] Compiling Main             ( tmp.hs, tmp.o )
Linking tmp ...
building '/nix/store/xcn490j6jsgji1mrgsvpvlx3pjjlpzqg-test-writers.drv'...
building '/nix/store/kx2l83gyvryka0x3nlkw8323kgs6qil3-test-writers.drv'...
building '/nix/store/hkpg0m44v3kvkgcjhzklbzhy7hyqbw1w-test_writers.drv'...
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-arch" "x86_64" "-L" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.0.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.1.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.2.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.3.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.4.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.5.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.6.rcgu.o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.tmp.7rcbfp3g-cgu.7.rcgu.o" "-o" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers" "/nix/store/6zvgl3gafnk8k16ksxfhzg2az9ars84n-test_writers.4i0hgzmnpf4h2pz.rcgu.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libstd-8ecebea30c961804.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-78b8371c8829edb9.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libobject-11c2250f76ee79de.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libaddr2line-138200afe481d69a.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libgimli-50edf7ed96dca7d7.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-f5add760ba32d8c5.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-107795e7ae43b1fc.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-1a28e1881ca37aeb.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libunwind-0b60cab5e184de42.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-20066734a370e7ce.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/liblibc-0688efbb4c57e170.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/liballoc-0c8da59b450dd54a.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-159337c2037ece59.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcore-1ba7a3965a87fb70.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-2f50c7e4a63f3402.rlib" "-lSystem" "-lresolv" "-lc" "-lm" "-liconv"
  = note: ld: library not found for -liconv
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

builder for '/nix/store/hkpg0m44v3kvkgcjhzklbzhy7hyqbw1w-test_writers.drv' failed with exit code 1
cannot build derivation '/nix/store/1laarw0v4xnl6jdancfswa9imzddad85-test-writers.drv': 1 dependencies couldn't be built
building '/nix/store/zmzl0phrj8fkzr8zch1xizysyybj7fbs-test_writers.drv'...
cannot build derivation '/nix/store/vl2imbpfjhqc8cxqidpk3kfdvvqgj8dy-test-writers.drv': 1 dependencies couldn't be built
error: build of '/nix/store/vl2imbpfjhqc8cxqidpk3kfdvvqgj8dy-test-writers.drv' failed
```

--- a/pkgs/development/compilers/ghc/8.10.4.nix
+++ b/pkgs/development/compilers/ghc/8.10.4.nix
@@ -93,6 +93,15 @@ let
   # see #84670 and #49071 for more background.
   useLdGold = targetPlatform.isLinux && !(targetPlatform.useLLVM or false) && !targetPlatform.isMusl;
 
+  runtimeDeps = [
+    targetPackages.stdenv.cc.bintools
+    coreutils
+  ]
+  # On darwin, we need unwrapped bintools as well (for otool)
+  ++ lib.optionals (stdenv.targetPlatform.linker == "cctools") [
+    targetPackages.stdenv.cc.bintools.bintools
+  ];
+
 in
 stdenv.mkDerivation (rec {
   version = "8.10.4";
@@ -235,7 +244,7 @@ stdenv.mkDerivation (rec {
     for i in "$out/bin/"*; do
       test ! -h $i || continue
       egrep --quiet '^#!' <(head -n 1 $i) || continue
-      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath [ targetPackages.stdenv.cc.bintools coreutils ]}"' $i
+      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath runtimeDeps}"' $i
     done
   '';
 

--- a/pkgs/development/compilers/ghc/8.8.4.nix
+++ b/pkgs/development/compilers/ghc/8.8.4.nix
@@ -102,6 +102,15 @@ let
   # see #84670 and #49071 for more background.
   useLdGold = targetPlatform.isLinux && !(targetPlatform.useLLVM or false) && !targetPlatform.isMusl;
 
+  runtimeDeps = [
+    targetPackages.stdenv.cc.bintools
+    coreutils
+  ]
+  # On darwin, we need unwrapped bintools as well (for otool)
+  ++ lib.optionals (stdenv.targetPlatform.linker == "cctools") [
+    targetPackages.stdenv.cc.bintools.bintools
+  ];
+
 in
 stdenv.mkDerivation (rec {
   version = "8.8.4";
@@ -242,7 +251,7 @@ stdenv.mkDerivation (rec {
     for i in "$out/bin/"*; do
       test ! -h $i || continue
       egrep --quiet '^#!' <(head -n 1 $i) || continue
-      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath [ targetPackages.stdenv.cc.bintools coreutils ]}"' $i
+      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath runtimeDeps}"' $i
     done
   '';
 

--- a/pkgs/development/compilers/ghc/9.0.1.nix
+++ b/pkgs/development/compilers/ghc/9.0.1.nix
@@ -94,6 +94,15 @@ let
   # see #84670 and #49071 for more background.
   useLdGold = targetPlatform.isLinux && !(targetPlatform.useLLVM or false) && !targetPlatform.isMusl;
 
+  runtimeDeps = [
+    targetPackages.stdenv.cc.bintools
+    coreutils
+  ]
+  # On darwin, we need unwrapped bintools as well (for otool)
+  ++ lib.optionals (stdenv.targetPlatform.linker == "cctools") [
+    targetPackages.stdenv.cc.bintools.bintools
+  ];
+
 in
 stdenv.mkDerivation (rec {
   version = "9.0.1";
@@ -225,7 +234,7 @@ stdenv.mkDerivation (rec {
     for i in "$out/bin/"*; do
       test ! -h $i || continue
       egrep --quiet '^#!' <(head -n 1 $i) || continue
-      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath [ targetPackages.stdenv.cc.bintools coreutils ]}"' $i
+      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath runtimeDeps}"' $i
     done
   '';
 

--- a/pkgs/development/compilers/ghc/head.nix
+++ b/pkgs/development/compilers/ghc/head.nix
@@ -104,6 +104,15 @@ let
   # see #84670 and #49071 for more background.
   useLdGold = targetPlatform.isLinux && !(targetPlatform.useLLVM or false) && !targetPlatform.isMusl;
 
+  runtimeDeps = [
+    targetPackages.stdenv.cc.bintools
+    coreutils
+  ]
+  # On darwin, we need unwrapped bintools as well (for otool)
+  ++ lib.optionals (stdenv.targetPlatform.linker == "cctools") [
+    targetPackages.stdenv.cc.bintools.bintools
+  ];
+
 in
 stdenv.mkDerivation (rec {
   inherit version;
@@ -242,7 +251,7 @@ stdenv.mkDerivation (rec {
     for i in "$out/bin/"*; do
       test ! -h $i || continue
       egrep --quiet '^#!' <(head -n 1 $i) || continue
-      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath [ targetPackages.stdenv.cc.bintools coreutils ]}"' $i
+      sed -i -e '2i export PATH="$PATH:${lib.makeBinPath runtimeDeps}"' $i
     done
   '';
 

